### PR TITLE
chore(deps): switched from `request` to `@theia/request`

### LINKED ISSUE #12401: Replace `request` with an alternative library
Replace [`request`](https://www.npmjs.com/package/request) with an alternative library. `request` is deprecated and causing `yarn audit` warnings. A list of alternatives is here: https://github.com/request/request/issues/3143#issue-427486416.

```
% yarn why request
yarn why v1.22.18
[1/4] 🤔  Why do we have the module "request"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "request@2.88.2"
info Reasons this module exists
   - "_project_#@theia#application-package" depends on it
   - Hoisted from "_project_#@theia#application-package#request"
   - Hoisted from "_project_#@theia#plugin-dev#request"
   - Hoisted from "_project_#@theia#plugin-ext-vscode#request"
   - Hoisted from "_project_#@theia#plugin-ext#request"
   - Hoisted from "_project_#jsdom#request"
info Disk size without dependencies: "696KB"
info Disk size with unique dependencies: "2.08MB"
info Disk size with transitive dependencies: "6.69MB"
info Number of shared dependencies: 37
✨  Done in 1.08s.
```

Since `jsdom` depends on an old `request` version, bumping `jsdom` would be also required:

```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Server-Side Request Forgery in Request                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ request                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ No patch available                                           │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ jsdom                                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ jsdom > request                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1091459                     │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

`jsdom` should be replaced anyway, so this would be a good opportunity to get rid of warnings:

```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Insufficient Granularity of Access Control in JSDom          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ jsdom                                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=16.5.0                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ jsdom                                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ jsdom                                                        │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1089185                     │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

Once `request` is gone and a more recent version of `jsdom` is used, one has to update the test helper here, and handle `'undefined'` property:

https://github.com/eclipse-theia/theia/blob/b56f399da8f6956b60c686033cd648496185c1cd/packages/core/src/browser/test/jsdom.ts#L51

### Feature Description:

AC: `yarn audit` should not generate warnings.

### PR DESCRIPTION
**Pull request by kittaakos**
#### What it does

This PR replaces  [`request`](https://www.npmjs.com/package/request) with [`@theia/request`](https://www.npmjs.com/package/@theia/request) to eliminate all dependency vulnerabilities detected by [`yarn audit`](https://classic.yarnpkg.com/lang/en/docs/cli/audit/).

```sh
% yarn audit
yarn audit v1.22.18
0 vulnerabilities found - Packages audited: 1912
✨  Done in 1.35s.
```

#### How to test

 - [ ] CI should pass, and
 - [ ] `yarn audit` should produce the `0 vulnerabilities found` result

Closes #12401

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

I have no idea how to test/verify the HTTP and GitHub plugin resolvers 😕 

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: dev-packages/application-package/src/npm-registry.ts [hunk 1]
```diff
@@ -150,19 +154,18 @@ export class NpmRegistry {
...
+        return new Promise(async (resolve, reject) => {
+            try {
+                const response = await this.request.request({ url, headers });
+                if (response.res.statusCode !== 200) {
+                    reject(new Error(`HTTP ${response.res.statusCode}: for ${url}`));
                 } else {
-                    const data = JSON.parse(body);
-                    resolve(data);
+                    const data = RequestContext.asJson(response);
+                    resolve(data as ViewResult);
```
Comment (msujew) [inline dev-packages/application-package/src/npm-registry.ts:164]: Minor:

```suggestion
                    const data = RequestContext.asJson<ViewResult>(response);
                    resolve(data);
```

**Comment by kittaakos** [comment]
Can somebody please advise what to do to fix the license check? Thank you!

```
X npm/npmjs/-/nwsapi/2.2.4, 
ERROR: Found results that aren't part of the baseline!
X npm/npmjs/@types/jsdom/21.1.1, 
```

File: packages/plugin-ext/src/main/node/plugin-http-resolver.ts [hunk 2]
```diff
@@ -74,10 +77,12 @@ export class HttpPluginDeployerResolver implements PluginDeployerResolver {
...
-            request.get(pluginResolverContext.getOriginId())
-                .on('error', err => {
-                    reject(err);
-                }).pipe(dest);
+            const stream = new Duplex();
+            this.request.request({ url: pluginResolverContext.getOriginId() }).then(({ buffer }) => {
+                stream.push(buffer);
+                stream.push(null);
+                stream.pipe(dest);
+            }, reject);
```
Comment (paul-marechal) [inline packages/plugin-ext/src/main/node/plugin-http-resolver.ts:85]:
```suggestion
        const buffer = await this.request.request({ url });
        dest.push(buffer);
        dest.end();
```

File: packages/plugin-ext/src/main/node/plugin-github-resolver.ts [hunk 3]
```diff
@@ -74,34 +78,30 @@ export class GithubPluginDeployerResolver implements PluginDeployerResolver {
...
+                    // parse redirect link
+                    const taggedValueArray = /^https:\/\/.*tag\/(.*)/gm.exec(redirectLocation);
+                    if (!taggedValueArray || taggedValueArray.length !== 2) {
+                        reject(new Error('The redirect link for latest is invalid ' + redirectLocation));
+                        return;
                     }
-                });
+
+                    // grab version of tag
+                    this.grabGithubFile(pluginResolverContext, orgName, repoName, filename, taggedValueArray[1], resolve, reject);
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/plugin-ext/src/main/node/plugin-github-resolver.ts]: OK. Thanks

File: dev-packages/application-package/src/npm-registry.ts [hunk 4]
```diff
@@ -140,30 +144,20 @@ export class NpmRegistry {
...
-                }
-            });
-        });
+        const response = await this.request.request({ url });
+        if (response.res.statusCode !== 200) {
+            throw new Error(`HTTP ${response.res.statusCode}: for ${url}`);
+        } else {
+            const data = RequestContext.asJson<ViewResult>(response);
+            return data;
+        }
```
Comment (paul-marechal) [inline dev-packages/application-package/src/npm-registry.ts:160]:
```suggestion
        }
        return RequestContext.asJson<ViewResult>(response);
```

**Review by paul-marechal — APPROVED** [review]
`yarn audit` no longer warns about vulnerabilities.

**Comment by paul-marechal** [comment]
@kittaakos there's not much you can do about new dependencies that cause dash-license to fail but @vince-fugnitto ran the tool in auto-review mode on your PR to make the check go green.

**Comment by kittaakos** [comment]
> @kittaakos there's not much you can do about new dependencies that cause dash-license to fail but @vince-fugnitto ran the tool in auto-review mode on your PR to make the check go green.

If the PR does not require a CQ, I would bump `jsdom` to `22+` instead of `17`. Let me know what you think. Thanks!

**Comment by vince-fugnitto** [comment]
> If the PR does not require a CQ, I would bump `jsdom` to `22+` instead of `17`. Let me know what you think. Thanks!

@kittaakos I'm fine with bumping to the latest if it works well, I'll perform a dash-licenses review as necessary.

**Comment by kittaakos** [comment]
> I'm fine with bumping to the latest if it works well, I'll perform a dash-licenses review as necessary.

I bump to the latest `jsdom` and push it. The tests pass. I push the changes in a few minutes. Thanks!

**Comment by paul-marechal** [comment]
I wanted to merge but GH tells me there are conflicts? Can you look into it? I'll merge then.

**Comment by kittaakos** [comment]
> Can you look into it? I'll merge then.

Thanks! I rebased the PR from the `main` branch and pushed the changes. I had to force-push it 😕

**Comment by kittaakos** [comment]
> Can you allow edits to your PR from collaborators

Done. Thanks for taking care of it. Great support from all of you 👏

**Pull request merged**